ConstExpr improvements for strings and moving towards arrays.

lattner opened the pull request:
Improve ConstExpr processing for strings to handle ones with literals correctly.
Start treating Builtin.RawPointer as addresses, which will be necessary for array
literals.

Refactor the code that analyzes calls to _allocateUninitializedArray and move it into
ConstExprEvaluator (not the best place perhaps, but better than what we had before).

Add a testcase showing that we handle empty and non-empty strings now.

lattner committed:
ConstExpr improvements for strings and moving towards arrays.

Improve ConstExpr processing for strings to handle ones with literals correctly.
Start treating Builtin.RawPointer as addresses, which will be necessary for array
literals.

Refactor the code that analyzes calls to _allocateUninitializedArray and move it into
ConstExprEvaluator (not the best place perhaps, but better than what we had before).

Add a testcase showing that we handle empty and non-empty strings now.

lattner commented:
@swift-ci please test tensorflow

mhong commented on lib/SILOptimizer/Mandatory/TFConstExpr.cpp:
```diff
@@ -1557,3 +1642,133 @@ computeConstantValues(ArrayRef<SILValue> values,
+/// Try to decode the specified apply of the _allocateUninitializedArray
+/// function in the standard library.  This attempts to figure out what the
+/// resulting elements will be.  This fills in the elements result and returns
+/// true on success.
```
some other existing functions return "true" on failure. 

My first preference is to "return true on success" consistently throughout the code base, but i understand that can involve changing large parts of the code base and may not be worthwhile.

My fallback preference is to "return true on failure" for consistency reasons.

What do you think?

rxwei replied:
+1

lattner replied:
Sure, will change.

mhong commented on lib/SILOptimizer/Mandatory/TFConstExpr.h:
```diff
@@ -61,6 +61,20 @@ class ConstExprEvaluator {
+  static bool decodeAllocUninitializedArray(ApplyInst *apply,
+                                            uint64_t numElements,
+                                            SmallVectorImpl<SILValue> &elements,
+                                   SmallPtrSet<SILInstruction*, 8> *arrayInsts);
```
align this param

rxwei replied:
I believe this is for fitting within 80 columns.

lattner replied:
Right, I flowed it a different way, it is a bit less weird.

mhong commented on lib/SILOptimizer/Mandatory/TFConstExpr.cpp:
```diff
@@ -838,16 +858,74 @@ ConstExprFunctionState::computeCallResult(ApplyInst *apply) {
+
+    auto v = SymbolicValue::getAggregate(elementConstants,
+                                         evaluator.getAllocator());
+    //llvm::errs() << "Got it: " << v;
```
remove this or wrap with DEBUG()?

lattner replied:
This is incomplete, I need to build this out when I have a way to represent arrays.  Still WIP.  I'll #if 0 is out :-)

mhong reviewed (approved):
Nice!

lattner commented:
Thanks for the feedback, I need to fix an issue handling StaticString that I regressed.  I'll try to do that later tonight or tomorrow.

lattner committed:
Address review feedback about parity of bool return.
Disable some tests using StaticString internally.  This patch
regresses them, I'll address it with greater refactorings down the
road.

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
@swift-ci please test tensorflow

lattner commented:
Oh, sorry @rxwei I missed your comment.  Please still do take a look, I'll incorporate any improvements into my next patch.  I need to change a lot of the internals... sigh.
